### ISSUE TITLE
Nil pointer when register instance to apiserver elb

### ISSUE THREAD

**Issue by jzhoucliqr** [kubernetes-sigs/cluster-api-provider-aws#1342]
/kind bug

- Cluster-api-provider-aws v0.4.4
- Kubernetes version: v1.15.3 

I have a AWSCluster which span across 3 AZs. During the bootstrap process, the subnetID info in AWSCluster.Spec.NetworkSpec is overwritten by a bug in my app. When control plane comes up, machine controller try to attach ELB, but AWSCluster.Spec.NetworkSpec.Subnets.IDs are missing, which cause nil pointer.

When pod crash and restarts, maybe because I have 3 AZs and cluster controller takes longer to finish reconcileSubnets, machine controller reaches this step earlier than cluster controller, most of the time it keeps crashing and never come back. Sometimes after a long time and hundreds of crashes, it finally come up.  

Logs

```
I1108 21:05:50.478078       1 routetables.go:55] controllers/AWSCluster "msg"="Subnet is already associated with route table" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123" "route-table-id"="rtb-054e42c1a76b7" "subnet-id"="subnet-de91843b76d7a"
I1108 21:05:50.478137       1 securitygroups.go:51] controllers/AWSCluster "msg"="Reconciling security groups" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:05:50.568015       1 network.go:53] controllers/AWSCluster "msg"="Reconcile network completed successfully" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:05:50.568078       1 bastion.go:45] controllers/AWSCluster "msg"="Reconciling bastion host" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:05:50.632078       1 awsmachine_controller.go:316] controllers/AWSMachine "msg"="Machine instance is running" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:05:50.680015       1 bastion.go:76] controllers/AWSCluster "msg"="Reconcile bastion completed successfully" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:05:50.680084       1 loadbalancer.go:42] controllers/AWSCluster "msg"="Reconciling load balancers" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:05:50.835409       1 loadbalancer.go:81] controllers/AWSCluster "msg"="Reconcile load balancers completed successfully" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:05:50.836223       1 controller.go:242] controller-runtime/controller "msg"="Successfully Reconciled"  "controller"="awscluster" "request"={"Namespace":"cluster-123","Name":"cluster-sample"}
I1108 21:05:50.895960       1 controller.go:242] controller-runtime/controller "msg"="Successfully Reconciled"  "controller"="awsmachine" "request"={"Namespace":"cluster-123","Name":"cluster-sample-controlplane-0-70fa2c2a"}
I1108 21:05:50.897721       1 awsmachine_controller.go:260] controllers/AWSMachine "msg"="Reconciling AWSMachine" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123"
I1108 21:05:50.897827       1 instances.go:81] controllers/AWSMachine "msg"="Looking for instance by id" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:05:51.053028       1 awsmachine_controller.go:316] controllers/AWSMachine "msg"="Machine instance is running" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:05:51.261976       1 controller.go:242] controller-runtime/controller "msg"="Successfully Reconciled"  "controller"="awsmachine" "request"={"Namespace":"cluster-123","Name":"cluster-sample-controlplane-0-70fa2c2a"}
I1108 21:05:51.262534       1 awsmachine_controller.go:260] controllers/AWSMachine "msg"="Reconciling AWSMachine" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123"
I1108 21:05:51.262628       1 instances.go:81] controllers/AWSMachine "msg"="Looking for instance by id" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:05:51.430678       1 awsmachine_controller.go:316] controllers/AWSMachine "msg"="Machine instance is running" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:05:51.642117       1 controller.go:242] controller-runtime/controller "msg"="Successfully Reconciled"  "controller"="awsmachine" "request"={"Namespace":"cluster-123","Name":"cluster-sample-controlplane-0-70fa2c2a"}
I1108 21:06:20.888459       1 awscluster_controller.go:130] controllers/AWSCluster "msg"="Reconciling AWSCluster" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:06:20.888565       1 network.go:21] controllers/AWSCluster "msg"="Reconciling network for cluster" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123" "cluster-name"="cluster-sample" "cluster-namespace"="cluster-123"
I1108 21:06:20.888621       1 vpc.go:42] controllers/AWSCluster "msg"="Reconciling VPC" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
I1108 21:06:20.888465       1 awsmachine_controller.go:260] controllers/AWSMachine "msg"="Reconciling AWSMachine" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123"
I1108 21:06:20.888789       1 instances.go:81] controllers/AWSMachine "msg"="Looking for instance by id" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:06:21.098313       1 awsmachine_controller.go:316] controllers/AWSMachine "msg"="Machine instance is running" "awsCluster"="cluster-sample" "awsMachine"="cluster-sample-controlplane-0-70fa2c2a" "cluster"="cluster-sample" "machine"="cluster-sample-controlplane-0-70fa2c2a" "namespace"="cluster-123" "instance-id"="i-1c4538f87"
I1108 21:06:21.143909       1 vpc.go:87] controllers/AWSCluster "msg"="Working on managed VPC" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123" "vpc-id"="vpc-0d05d0df9c84bf831"
I1108 21:06:21.143992       1 subnets.go:44] controllers/AWSCluster "msg"="Reconciling subnets" "awsCluster"="cluster-sample" "cluster"="cluster-sample" "namespace"="cluster-123"
E1108 21:06:21.328200       1 runtime.go:73] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 321 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1870b40, 0x2e23750)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190817020851-f2f3a405f61d/pkg/util/runtime/runtime.go:51 +0x82
panic(0x1870b40, 0x2e23750)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/services/elb.(*Service).RegisterInstanceWithAPIServerELB(0xc000b55788, 0xc00013e000, 0x1aeb284, 0x1e)
	/workspace/pkg/cloud/services/elb/loadbalancer.go:147 +0x1a2
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).reconcileLBAttachment(0xc0007153b0, 0xc0005a10e0, 0xc0001844d0, 0xc00013e000, 0x2, 0x2)
	/workspace/controllers/awsmachine_controller.go:372 +0xb9
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).reconcileNormal(0xc0007153b0, 0x1dfe6c0, 0xc0000bc068, 0xc0005a10e0, 0xc0001844d0, 0x0, 0x1e1d0e0, 0xc000774b40, 0x1e0f8e0)
	/workspace/controllers/awsmachine_controller.go:325 +0x6ff
sigs.k8s.io/cluster-api-provider-aws/controllers.(*AWSMachineReconciler).Reconcile(0xc0007153b0, 0xc0007f7710, 0xb, 0xc0004e39b0, 0x2d, 0x2e3c300, 0x0, 0x0, 0x0)
	/workspace/controllers/awsmachine_controller.go:144 +0xa62

```

**Comment by jzhoucliqr** [comment]
Figured out that the subnetID overwritten is because of a bug in my app.  The fix in #1342 checks for nil pointer is still helpful.

**Comment by jzhoucliqr** [comment]
already fixed in #1341
